### PR TITLE
core(full-page-screenshot): add full page screenshot to artifacts

### PR DESCRIPTION
**Pull request by mattzeunert**
Capture a screenshot of the entire page, rather than just the above the fold content.

Splitting this off from [this PR](https://github.com/GoogleChrome/lighthouse/pull/8797) which actually uses the artifact.

What's the best way to test this? Add a new smoke test and unit test the gatherer with a mocked driver?

**Related Issues/PRs**

Part of https://github.com/GoogleChrome/lighthouse/issues/7327

### CODE REVIEW

File: lighthouse-core/gather/driver.js [hunk 1]
```diff
@@ -1423,13 +1423,14 @@ class Driver {
...
    * @return {Promise<void>}
    */
-  async beginEmulation(settings) {
+  async beginEmulation(settings, deviceMetricOverrides) {
     if (settings.emulatedFormFactor === 'mobile') {
-      await emulation.enableNexus5X(this);
+      await emulation.enableNexus5X(this, deviceMetricOverrides);
     } else if (settings.emulatedFormFactor === 'desktop') {
-      await emulation.enableDesktop(this);
+      await emulation.enableDesktop(this, deviceMetricOverrides);
```
Comment (evenstensberg) [inline lighthouse-core/gather/driver.js]: Would it be appropriate to hook the options on the driver itself?

Comment (mattzeunert) [reply lighthouse-core/gather/driver.js]: What do you mean by that?

Comment (evenstensberg) [reply lighthouse-core/gather/driver.js]: If the extra param could be removed in favor for having it as a prop on `this`

Comment (patrickhulce) [reply lighthouse-core/gather/driver.js]: I'm not sure we want to clutter driver with additional properties just for the method here. It's a pretty busy class already 😆

File: lighthouse-core/gather/gatherers/full-page-screenshot.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+class FullPageScreenshot extends Gatherer {
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   * @param {number} maxScreenshotHeight
+   * @return {Promise<LH.Artifacts.FullPageScreenshot>}
+   */
+  async _takeScreenshot(passContext, maxScreenshotHeight) {
+    const driver = passContext.driver;
+    const metrics = await driver.sendCommand('Page.getLayoutMetrics');
+    const width = await driver.evaluateAsync(`window.innerWidth`) * 2;
```
[the remark this answers is not in the document]
Comment (mattzeunert) [reply lighthouse-core/gather/gatherers/full-page-screenshot.js]: Whoops, that slipped in when testing large screenshots.

File: lighthouse-core/gather/gatherers/full-page-screenshot.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    const data = 'data:image/jpeg;base64,' + result.data;
+
+    // Revert resized page
+    await driver.beginEmulation(passContext.settings);
+
+    return {
+      data,
+      width,
+      height,
+      // todo: do we need these? should we create a new type? should we make them optional on the existing type? should we try to provide accurate values?
```
[the remark this answers is not in the document]
Comment (mattzeunert) [reply lighthouse-core/gather/gatherers/full-page-screenshot.js]: Sounds good. It makes sense to have the timings for the filmstrip, but Filmstrip is a separate type from Screenshot.

**Comment by connorjclark** [comment]
we did this in #10016

**Pull request closed without merging**